### ISSUE TITLE
multiprocessing.joinablequeue error when running demo code in jupyter lab

### ISSUE THREAD

**Issue by drhades** [alphatwirl/mantichora#4]
Jupyter Lab: 1.2.6
Python: 3.8.0

Running the following demo code:
<img width="918" alt="截屏2020-03-1118 17 22" src="https://user-images.githubusercontent.com/82670/76406391-9acbaa80-63c4-11ea-8bc4-a68c3d9fe54b.png">

Background Jupyter service error logging (and the above cell sinked in running status, cannot be stopped):

Process Worker-2:
Process Worker-1:
Process Worker-3:
Traceback (most recent call last):
Traceback (most recent call last):
  File "~/.pyenv/versions/3.8.0/lib/python3.8/multiprocessing/process.py", line 313, in _bootstrap
    self.run()
  File "~/.pyenv/versions/3.8.0/lib/python3.8/site-packages/mantichora/hubmp.py", line 235, in run
    self._run_tasks()
  File "~/.pyenv/versions/3.8.0/lib/python3.8/multiprocessing/process.py", line 313, in _bootstrap
    self.run()
  File "~/.pyenv/versions/3.8.0/lib/python3.8/site-packages/mantichora/hubmp.py", line 250, in _run_tasks
    message = self.task_queue.get()
  File "~/.pyenv/versions/3.8.0/lib/python3.8/multiprocessing/queues.py", line 116, in get
    return _ForkingPickler.loads(res)
  File "~/.pyenv/versions/3.8.0/lib/python3.8/site-packages/mantichora/hubmp.py", line 235, in run
    self._run_tasks()
  File "~/.pyenv/versions/3.8.0/lib/python3.8/site-packages/mantichora/hubmp.py", line 250, in _run_tasks
    message = self.task_queue.get()
  File "~/.pyenv/versions/3.8.0AttributeError: Can't get attribute 'task_loop' on <module '__main__' (built-in)>
Traceback (most recent call last):
/lib/python3.8/multiprocessing/queues.py", line 116, in get
    return _ForkingPickler.loads(res)
  File "~/.pyenv/versions/3.8.0/lib/python3.8/multiprocessing/process.py", line 313, in _bootstrap
    self.run()
  File "~/.pyenv/versions/3.8.0/lib/python3.8/site-packages/mantichora/hubmp.py", line 235, in run
    self._run_tasks()
  File "~/.pyenv/versions/3.8.0/lib/python3.8/site-packages/mantichora/hubmp.py", line 250, in _run_tasks
    message = self.task_queue.get()
AttributeError: Can't get attribute 'task_loop' on <module '__main__' (built-in)>
  File "~/.pyenv/versions/3.8.0/lib/python3.8/multiprocessing/queues.py", line 116, in get
    return _ForkingPickler.loads(res)
AttributeError: Can't get attribute 'task_loop' on <module '__main__' (built-in)>

**Comment by TaiSakuma** [comment]
Thank you for letting me know. I'll have a look.

**Comment by drhades** [comment]
For reference, 
defining task_loop() in another python file and importing, it'll be ok.

**Comment by TaiSakuma** [comment]
Yes, I'm just working on this now. The code in [this branch](https://github.com/alphatwirl/mantichora/tree/dev-20200314-01-py38-startmethod) might work.

**Comment by TaiSakuma** [comment]
I released the version [0.9.8](https://pypi.org/project/mantichora/0.9.8/).

It works fine in Python 3.8 on Jupyter Notebook in my environment. 

<img width="759" alt="Screen Shot 2020-03-14 at 11 42 00 AM" src="https://user-images.githubusercontent.com/1388081/76685317-14dc7900-65e9-11ea-990f-99191d7f3852.png">

Please let me know if the problem persists.

The cause of the problem appears to be the change in the default start method in multiprocessing in Python 3.8. The default start method was "fork" until Python 3.7.  But it changed to "spawn" in Python 3.8. The spawn method has more restrictions on how the main module is written as described in [this section of the document for mulriprocessing](https://docs.python.org/3/library/multiprocessing.html#the-spawn-and-forkserver-start-methods). The mantichora version [0.9.8](https://pypi.org/project/mantichora/0.9.8/) uses the "fork" mode.

**Comment by drhades** [comment]
@TaiSakuma 
Thanks for your quick fix and detail explaination!

After upgrading to 0.9.8, the above demo code work well,
but I get the following error:
![image](https://user-images.githubusercontent.com/82670/76703088-b955dc80-6709-11ea-8558-e59341dff56b.png)

when using mantichora background, code like:   
```python
def task_func(*args)
    for i in atpbar(...):
        do some network things
    return something

def parallel_func()
    split something into K parts
    task_partial = partial(task_func, ...)
    with mantichora() as mcore:
        for i in range(K):
            mcore.run(task_partial, ...)
        mcore.receive_all()
``` 
It works fine by 0.9.7.

**Comment by TaiSakuma** [comment]
Thank you for letting me know. I will look into it.

**Comment by TaiSakuma** [comment]
I released mantichora version [0.9.9](https://pypi.org/project/mantichora/0.9.9/).

I added the option `mp_start_method` to `mantichora()` so that uses can choose the start method of multiprocessing. The possible values are `fork`, `spawn`, and `forkserver`. The default is `fork`. For example, if you would like to use the spawn method, which was used in mantichora v0.9.7 in Python 3.8, you can do so as

```python
mantichora(mp_start_method='spawn')
```

The `spawn` and `forkserver` mothods have more restrictions, which are already mentioned above in this issue.

If you need to use the `fork` method, you might get the same error of `may have been in progress in another thread when fork() was called`. If you are getting the same error and you are on MacOS, setting the environment variable `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` might solve the problem as suggested at StackOverflow [here](https://stackoverflow.com/questions/50168647/multiprocessing-causes-python-to-crash-and-gives-an-error-may-have-been-in-progr), which you might have found already.

Please let me know if this doesn't solve the problem or causes other problems.

Thank you.

**Comment by TaiSakuma** [comment]
I assume that the issue has been resolved. Please create a new issue if not.
